### ISSUE TITLE
Proposal: more granularity of control over accessibility levels

### ISSUE THREAD

**Issue by nxrighthere** [dotnet/csharplang#3305]
## Problem
The current modifiers are lack granularity of control over accessibility levels relative to a parent namespace where (for instance) implementation of a class that carries public functionality for the end-user through derived classes, can be used for inheritance within a parent namespace but should be forbidden for that in foreign namespaces by design.

## Example
A base component:
```csharp
/// <summary>
/// The base class of components that has a transform and supports attachment
/// </summary>
public abstract class Component : IEquatable<Component> {
    private IntPtr pointer;

    internal IntPtr Pointer {
        get {
            if (!IsCreated)
                throw new InvalidOperationException("Transformable component is not created");

            return pointer;
        }

        set {
            pointer = value;
        }
    }

    /// <summary>
    /// The actor which attached the component
    /// </summary>
    public Actor Actor { get; internal set; }

    /// <summary>
    /// Returns <c>true</c> if the component is created
    /// </summary>
    public bool IsCreated {
        get => pointer != IntPtr.Zero && Object.isValid(pointer) && Actor.IsSpawned;
    }

    ...
}
```
One of the derived components:
```csharp
/// <summary>
/// A component that can be transformed or attached, but has no rendering or collision capabilities
/// </summary>
public sealed class SceneComponent : Component, ISceneComponent {
    ...
    
    /// <summary>
    /// Attaches the component to another component, optionally at a named socket
    /// </summary>
    public bool AttachToComponent(Component parent, AttachmentTransformRule attachmentRule, string socketName = null) {
        if (parent == null)
            throw new ArgumentNullException(nameof(parent));

        return attachToComponent(Pointer, parent.Pointer, attachmentRule, socketName);
    }

    ... 
}
```
The `Component` class represents a base implementation of a component in a game engine with base functionality that can be used by the end-user, but it should be forbidden for inheritance in the user-code within foreign namespaces because it carries engine-specific internal data, but with public functionality that has a constrained implementation which exlustively belongs to a framework of the game engine.

## Solution
A new `constrained` modifier similar to `abstract` modifier, but instead indicates that a class can't be instantiated and can be inherited only within a parent namespace. This also can be applied to interfaces with similar requirements.

With proposed modifier, the base class from the example becomes:
```csharp
public constrained class Component : IEquatable<Component> {
    ... 
}
```

**Comment by HaloFour** [comment]
You separate library code from user code through assembly boundaries, not namespaces.  There is no runtime support for enforcing any kind of accessibility modifiers.  Namespaces don't really exist at runtime.

**Comment by nxrighthere** [comment]
This should be a compilation-time feature that has nothing to do with the runtime. It explicitly indicates to the end-user that a particular class can't inherited in foreign namespace.

**Comment by nxrighthere** [comment]
This should be a compilation-time feature that has nothing to do with the runtime. It explicitly indicates to the end-user that a particular class can't be inherited in a foreign namespace.

**Comment by HaloFour** [comment]
This sounds like something that can be enforced by a custom analyzer applied to the project.  You'd mark the base class or interface that would be "constrained" by an attribute and the analyzer would emit an error if any subclasses/implementors of that type were declared in a different namespace.

**Comment by nxrighthere** [comment]
The problem with a custom analyzer approach is that the framework is distributed as an assembly which is ready for consumption by the end-user without any external effort. We can't simply ask a customer to do this on his end, and we don't have control over his compilation pipeline in this case.

**Comment by HaloFour** [comment]
If the end user is defining their own types in their own assemblies deriving from types in your assembly I fail to see why you care what namespace they happen to define them in.

**Comment by nxrighthere** [comment]
It's about to prevent misconceptions like a user assumes that a class allows inheritance, and intuitively can be used for the implementation of custom components derived from the base component of the engine, while it's not.

**Comment by CyrusNajmabadi** [comment]
> It's about to prevent misconceptions like a user assumes that a class allows inheritance,

Couldn't this be addressed by not allowing your class to be inherited?

**Comment by nxrighthere** [comment]
Unfortunately not... We were thinking about redesigning and refactoring code where inheritance is not a thing at all, but it overcomplicates logic of the framework significantly and leads to code duplication. We also need to preserve a pattern matching based on the base component across derived classes.

**Comment by HaloFour** [comment]
You can prevent inheritance from outside of your assembly by making the constructor `internal`.

**Comment by nxrighthere** [comment]
The base class is an abstract public class with public functionality, there are no constructors. It can't be restricted to internal usage because the base class carries the generic public functionality for the end-user that should remain accessible in derived sealed classes.

**Comment by HaloFour** [comment]
Then how does the end user create/obtain an instance of this class?

**Comment by nxrighthere** [comment]
A user instantiate derived classes that inherit the base class.

**Comment by HaloFour** [comment]
So then add an `internal` constructor to `Component`.  That will allow you to extend that class only within your assembly.  But the derived `sealed` classes can have public constructors, so that consumers from outside assemblies can instantiate them directly.

**Comment by nxrighthere** [comment]
As I already said, `Component` is an abstract class, there's no need for any constructors, it's not about the base class initialization in one or another way.

What you are suggesting is not preventing the user from inheriting the base class and let him assume that a custom component can be implemented based on it.

**Comment by theunrepentantgeek** [comment]
@nxrighthere wrote

> This should be a compilation-time feature that has nothing to do with the runtime. 

If this approach was taken, you'd have a constraint only enforced by the new C# compiler that would be entirely ignored by older C# compilers, VN.NET, F# and any other language that targets the CLR. That's not very constrained at all.

> A new constrained modifier similar to abstract modifier, but instead indicates that a class can't be instantiated and can be inherited only within a parent namespace.

There's no relationship between the name of an assembly and the namespace(s) contained. Namespaces are purely a C# construct that doesn't exist at runtime - and it's perfectly legal to do something like this:

``` csharp
using System;

namespace System
{
    public struct Sneaky {
        public void M() {
        }
    }
}
```

If the **constrained** modifier was enforced by namespace, any consumer of your class could use this approach to trivially bypass it.

Why not use the features already present in the language to achieve what you want, as suggested by @HaloFour? If you create a **public** **abstract** class with an **internal** constructor, even an empty one, the only permitted descendants are in the current assembly. 

``` csharp
public abstract class Component
{
    // Prevent direct inheritance outside this assembly
    internal Component() { }

}
```

I've used this technique myself. It works.

Your leaf classes will need to be marked **sealed** as well, but it appears that you're already doing that.

**Comment by HaloFour** [comment]
@nxrighthere 

> As I already said, `Component` is an abstract class, there's no need for any constructors, it's not about the base class initialization in one or another way.

All derived classes are required to call the constructor of the base class, even if it's abstract.  If you don't define one yourself the compiler automatically adds one that is `public.  So the net effect of adding an `internal` constructor to the abstract class is that only classes defined within the same assembly can extend from that abstract class.

**Comment by nxrighthere** [comment]
> If this approach was taken, you'd have a constraint only enforced by the new C# compiler that would be entirely ignored by older C# compilers, VB.NET, F# and any other language that targets the CLR. That's not very constrained at all.

In our case, we have an embedded .NET Core host into a game engine, so there's only one way to compile user-code in accordance with its version. Other cases should be properly investigated and taken into account. no doubt.

> If the constrained modifier was enforced by namespace, any consumer of your class could use this approach to trivially bypass it.

Our concern is not that a user can bypass this, but we can explicitly indicate and tell him that he's going the wrong path. With reflection, users can bypass any language-specific limitations, and this is where we don't take any responsibilities.

> I've used this technique myself. It works.

This might be an option, going to investigate this quickly, thanks.

**Comment by nxrighthere** [comment]
@HaloFour This might be a good temporary solution, investigating this now, thanks.

**Comment by nxrighthere** [comment]
@theunrepentantgeek @HaloFour It works perfectly, thanks for the input once again. While this feels like a hack triggering `CS1729`, it's still better than nothing.

At this point, I would like to hear any toughs from people behind the language, please. @gafter

**Comment by CyrusNajmabadi** [comment]
> While this feels like a hack triggering CS1729 in the user-code, 

This is not a hack.  It's entirely appropriate and how the language is intended to work for this scenario.  If you do not want external subclasses of your components, you make your constructor internal.   :)

**Comment by CyrusNajmabadi** [comment]
> I've used this technique myself. It works.

Roslyn also does this in its own code all over the place.  It's a totally normal and standard way to design api hierarchies with inheritance that you want to expose but not allow external derivations from.

**Comment by CyrusNajmabadi** [comment]
> Unfortunately not... We were thinking about redesigning and refactoring code where inheritance is not a thing at all

I wasn't say to not use inheritance.  I was saying to not allow *your* types to be inherited from :)

**Comment by nxrighthere** [comment]
I'm not sure how it's not a hack if the compiler pointing to a wrong error and wrong assumptions instead of a solid reason and explanation?

**Comment by jnm2** [comment]
A hack is use of a feature in a manner that subverts its intended purpose. I don't see any reason to think that internal constructors on abstract classes are subverting the intended purpose of visibility modifiers.

**Comment by CyrusNajmabadi** [comment]
> which causing wrong assumptions instead of a solid reason and explanation?

It's a solid reason and explanation: You cannot subclass this because i don't expose any constructors for you to use.  Subclassing is only allowed `internal`ly

**Comment by CyrusNajmabadi** [comment]
>  I don't see any reason to think that internal constructors on abstract classes are subverting the intended purpose of visibility modifiers. It's the canonical way to say "I'm allowed to inherit from this, but you're not."

Exactly.  This combination is 100% supported, expected, and useful for this exact case.  It's idiomatic and used commonly.  There's no need for another feature which simply duplicates what already exists.
